identity_stardust/src/document/stardust_document.rs

```diff
@@ -0,0 +1,753 @@
+
+use crate::document::stardust_document_metadata::StardustDocumentMetadata;
+use crate::error::Result;
+use crate::state_metadata::StateMetadataEncoding;
+use crate::state_metadata::PLACEHOLDER_DID;
+use crate::StateMetadataDocument;
+
+// TODO: replace with StardustDID struct.
+type StardustDID = CoreDID;
+
+/// A DID URL conforming to the IOTA DID method specification.
+///
+/// See [`DIDUrl`].
+// TODO: move to `did` module.
+pub type StardustDIDUrl = DIDUrl<StardustDID>;
+
+/// A [`VerificationMethod`] adhering to the IOTA DID method specification.
+pub type StardustVerificationMethod = VerificationMethod<StardustDID, Object>;
+
+/// A [`Service`] adhering to the IOTA DID method specification.
+pub type StardustService = Service<StardustDID, Object>;
+
+/// A [`CoreDocument`] whose fields adhere to the IOTA DID method specification.
+pub type StardustCoreDocument = CoreDocument<StardustDID>;
+
+/// A DID Document adhering to the IOTA DID method specification.
+///
+/// This extends [`CoreDocument`].
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct StardustDocument {
+  #[serde(rename = "doc")]
+  pub(crate) document: StardustCoreDocument,
+  #[serde(rename = "meta")]
+  pub metadata: StardustDocumentMetadata,
+}
+
+impl StardustDocument {
+  // ===========================================================================
+  // Constructors
+  // ===========================================================================
+
+  /// Constructs an empty DID Document with a [`StardustDocument::placeholder_did`] identifier.
+  // TODO: always take Option<NetworkName> or `new_with_options` for a particular network?
+  pub fn new() -> Self {
+    Self::new_with_id(Self::placeholder_did().clone())
+  }
+
+  /// Constructs an empty DID Document with the given identifier.
+  pub fn new_with_id(id: StardustDID) -> Self {
+    // PANIC: constructing an empty DID Document is infallible, caught by tests otherwise.
+    let document: StardustCoreDocument = CoreDocument::builder(Object::default())
+      .id(id)
+      .build()
+      .expect("empty StardustDocument constructor failed");
+    let metadata: StardustDocumentMetadata = StardustDocumentMetadata::new();
+    Self { document, metadata }
+  }
+
+  // ===========================================================================
+  // Properties
+  // ===========================================================================
+
+  /// Returns the placeholder DID of newly constructed DID Documents,
+  /// `"did:0:0"`.
+  // TODO: generalise to take network name?
+  pub fn placeholder_did() -> &'static StardustDID {
+    &PLACEHOLDER_DID
+  }
+
+  /// Returns the DID document identifier.
+  pub fn id(&self) -> &StardustDID {
+    self.document.id()
+  }
+
+  /// Returns a reference to the DID controllers.
+  ///
+  /// NOTE: controllers are determined by the `state_controller` unlock condition of the output
+  /// during resolution and are omitted when publishing.
```

Comment:
It's not, it's an explanation as to why changes such as adding/removing controllers on the `StardustDocument` itself will not persist through publication. Though this is not implemented yet.

Reply:
Then wouldn't it be enough to mention it in those places where one would modify the controllers? Which only seems to be `core_document_mut`, since `controller_mut` has (deliberately, I assume) been omitted.

Reply:
Because it also mentions the `state_controller` connection, and it's more relevant to `controller` than `core_document_mut` (which is not exposed to the Wasm bindings).